src/circuit/circuit_builder.rs

```diff
@@ -892,19 +886,39 @@ where
         self.inner().parent.clone()
     }
 
-    /// Returns a reference to the cache associated with the circuit.
+    /// Lookup a value in the circuit cache or insert a value on cache miss.
     ///
     /// See [`cache`] module documentation for details.
-    pub(crate) fn cache(&self) -> RefMut<'_, CircuitCache> {
-        let circuit = self.inner_mut();
+    pub(crate) fn cache_lookup_or_insert_with<K, F>(&self, key: K, mut f: F) -> RefMut<'_, K::Value>
+    where
+        K: 'static + TypedMapKey<CircuitStoreMarker>,
+        F: FnMut() -> K::Value,
+    {
+        // Don't use `store.entry()`, since `f` may need to perform
+        // its own cache lookup.
+        if self.inner().store.contains_key(&key) {
+            return RefMut::map(self.inner_mut(), |c| c.store.get_mut(&key).unwrap());
+        }
```

Comment:
This causes two lookups in the happy path which is less than ideal. There's two ways to address this, one would be using hashbrown to allow dancing around the hashing somewhat and the other would be to use [`RefMut::filter_map()`](https://github.com/rust-lang/rust/issues/81061) which doesn't seem to be stable yet which is very unfortunate.
Ideally we'd do both, so I'll just put that as a todo
```suggestion
        // its own cache lookup.
        // TODO: Only perform hashing once by using hashbrown's raw api
        // TODO: Use `RefMut::filter_map()` to only perform one lookup in the happy path
        //       https://github.com/rust-lang/rust/issues/81061
        if self.inner().store.contains_key(&key) {
            return RefMut::map(self.inner_mut(), |c| c.store.get_mut(&key).unwrap());
        }
```

Reply:
This API is used at circuit construction time. I don't care about the overhead of extra lookup.  `filter_map` would be nice, but raw API isn't worth it.